### PR TITLE
feat!: add simpler Option style

### LINKED ISSUE #470: Add support to Option<T>
## Description

Today if a contract uses the sway object `Option<T>` as a return or argument of a function, we parse it as a custom Enum from the ABI, requiring the user to take action and pass an object `{ None: [] }` or `{ Some: <value> }` to match the custom enum.

The ideal implementation of `Option<T>` on typescript would be to abstract the `enum Option` and;
- In case of argument: simplify the argument to be the valid value (`T`) or optional in this case null or undefined.
- On the return: we should return the value `T` or null/undefined.

Todos;
- [ ] Implement an abstraction for the `Option<T>` by identifying it from the json ABI.
- [ ] Add the abstraction type inside the `typechain` to match the implementation.

### Contract

```
contract;

use std::{option::Option};

abi MyContract {
  fn test_option(param: Option<u32>) -> Option<u32>;
}

impl MyContract for Contract {
  fn test_option(param: Option<u32>) -> Option<u32> {
    param
  }
}
```

### Current way on Typescript

```ts
// When calling the contract, we require the user to 
// create an object { None: [] } for no params and { Some: 12 }
// when it will provide the value.
const { value } = await contract.functions
  .test_option({
    None: [],
  })
  .call();
// value will be {  None: [] }
const { value } = await contract.functions
  .test_option({
    Some: 12,
  })
  .call();
// value will be {  Some: 12 }
```

### Desired implementation on Typescript

```ts
// When calling the contract that has a params Option<T>, 
// we can make the param optional and create the underlining abstraction
// to create the correct params object { None: [] } or { Some: 12 }
const { value } = await contract.functions
  .test_option()
  .call();
// value will be null
const { value } = await contract.functions
  .test_option(12)
  .call();
// value will be 12
```

### PR DESCRIPTION
**Pull request by camsjams**
Closes #470

- Adds Option support
- Adds tests for Option
- Updates Typechain types to use new Option style

### CODE REVIEW

File: packages/contract/src/__test__/coverage-contract/coverage-contract.test.ts [hunk 1]
```diff
@@ -175,4 +177,24 @@ describe('Coverage Contract', () => {
...
+  it('should test Option<u32> extraction', async () => {
+    const INPUT_SOME = 123;
+    const { value: Some } = await contractInstance.functions
+      .echo_option_extract_u32(INPUT_SOME)
+      .call();
+    expect(Some).toStrictEqual(INPUT_SOME);
+
+    const INPUT_NONE = undefined;
+    const { value: None } = await contractInstance.functions
+      .echo_option_extract_u32(INPUT_NONE)
```
Comment (luizstacio) [inline packages/contract/src/__test__/coverage-contract/coverage-contract.test.ts:196]: On my opinion it should be possible to execute this `echo_option_extract_u32()`

**Review by luizstacio — CHANGES_REQUESTED** [review]
I think we should also include typechain on this change.

File: packages/abi-coder/src/abi-coder.ts [hunk 2]
```diff
@@ -82,7 +83,9 @@ export default class AbiCoder {
         obj[component.name] = this.getCoder(component);
         return obj;
       }, {});
-      return new EnumCoder(enumMatch.name, coders);
+      return param.type === 'enum Option'
```
Comment (QuinnLee) [inline packages/abi-coder/src/abi-coder.ts:86]: I was trying to think of a better way to de-nest this but couldn't.  My gut tells me we should refactor `getCoder` but probably not a blocker for this pr

File: packages/abi-coder/src/abi-coder.ts [hunk 3]
```diff
@@ -82,7 +84,9 @@ export default class AbiCoder {
         obj[component.name] = this.getCoder(component);
         return obj;
       }, {});
-      return new EnumCoder(enumMatch.name, coders);
+      return param.type === OPTION_CODER_TYPE
+        ? new OptionCoder(enumMatch.name, coders)
+        : new EnumCoder(enumMatch.name, coders);
```
Comment (luizstacio) [inline packages/abi-coder/src/abi-coder.ts:89]:
```suggestion
     const isOptionEnum = param.type === OPTION_CODER_TYPE;
     if (isOptionEnum) {
         return new OptionCoder(enumMatch.name, coders);
     }
     return new EnumCoder(enumMatch.name, coders);
```

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by luizstacio — APPROVED** [review]
Just one code comment.

**Review by luizstacio — APPROVED** [review]

**Review by QuinnLee — APPROVED** [review]

**Pull request merged**